## 🧮 Derivable Judgement


### 📚 Importing Libraries:- 


In [115]:
import pandas as pd
import numpy as np
from scipy.stats import chi2_contingency, chi2
from scipy.stats import f
from scipy.stats import t
from scipy.stats import chi2

print("Libraries Imported succesfully !")


Libraries Imported succesfully !


### 📂 Load the dataset :- 


In [79]:
df = pd.read_excel("Health_Dataset_500.xlsx")
df.head()

,record_id,age_group,age,weight,gender,region,smoking_status,exercise_frequency,bmi,blood_pressure,diabetes,hypertension,cholesterol_level,glucose_level,visit_date
0,3c9589cf-ba6d-452a-915d-390640f71b3f,60+,62,69,Male,East,Former Smoker,Rarely,27.7,175.7,False,False,275.9,177.1,2025-01-16
1,ebf17944-a083-4aed-a919-532c92bb6c1d,46-60,48,100,Female,West,Smoker,Rarely,28.8,128.6,False,False,133.3,106.3,2024-08-07
2,b184e763-e038-4f5e-be19-bbf7a38e240d,60+,81,83,Female,East,Non-Smoker,Never,25.5,128.9,True,True,293.5,108.6,2025-03-02
3,9044384c-20d2-4c00-8403-8ab699825ecd,60+,79,98,Female,West,Smoker,Daily,26.1,108.7,False,False,132.3,91.0,2024-03-06
4,c0c80ba7-6c24-447e-9dfd-5a81093c77cd,60+,76,48,Other,East,Former Smoker,Never,20.7,179.0,True,False,146.4,96.4,2025-07-17


In [80]:
df.tail()

,record_id,age_group,age,weight,gender,region,smoking_status,exercise_frequency,bmi,blood_pressure,diabetes,hypertension,cholesterol_level,glucose_level,visit_date
495,fa515e8b-1b26-4226-8f27-0810738e17fe,18-25,22,75,Other,East,Non-Smoker,Rarely,30.7,133.4,False,False,277.2,96.0,2024-09-15
496,8ab0cc81-4d5c-4f58-a5aa-05b10fb83bb1,46-60,49,50,Male,East,Non-Smoker,Rarely,33.5,160.3,False,True,160.3,102.5,2025-07-30
497,9e980cce-156c-400b-8a80-ec33b1f3eeb6,18-25,20,52,Other,West,Smoker,Daily,23.9,161.5,True,True,121.1,85.3,2024-01-01
498,1b451192-d3b2-4582-b0cd-e937848e88c5,26-35,33,63,Male,West,Non-Smoker,Never,18.7,116.8,False,True,234.2,108.8,2024-09-03
499,b799a995-a74a-48d3-a93e-07834a799f43,36-45,38,90,Female,North,Smoker,Weekly,21.2,151.5,True,True,132.2,116.7,2024-10-21


In [81]:
df.describe()

,age,weight,bmi,blood_pressure,cholesterol_level,glucose_level
count,500.000000,500.000000,500.000000,500.000000,500.000000,500.000000
mean,42.674000,77.506000,26.534600,137.843200,211.479400,147.347400
std,18.021801,19.394594,4.850133,25.182546,50.259852,41.866909
min,18.000000,45.000000,18.000000,95.000000,121.100000,70.100000
25%,28.000000,60.000000,22.400000,116.400000,169.100000,110.375000
50%,39.000000,78.000000,26.400000,137.100000,210.700000,150.450000
75%,55.250000,95.000000,30.850000,161.250000,253.400000,180.125000
max,85.000000,110.000000,34.900000,179.600000,299.800000,219.900000


In [82]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   record_id           500 non-null    object 
 1   age_group           500 non-null    object 
 2   age                 500 non-null    int64  
 3   weight              500 non-null    int64  
 4   gender              500 non-null    object 
 5   region              500 non-null    object 
 6   smoking_status      500 non-null    object 
 7   exercise_frequency  500 non-null    object 
 8   bmi                 500 non-null    float64
 9   blood_pressure      500 non-null    float64
 10  diabetes            500 non-null    bool   
 11  hypertension        500 non-null    bool   
 12  cholesterol_level   500 non-null    float64
 13  glucose_level       500 non-null    float64
 14  visit_date          500 non-null    object 
dtypes: bool(2), float64(4), int64(2), object(7)
memory usage:

### 🤔 Q1. Formulate at least two hypotheses from the dataset. Example:
* H₀: 🚬 Smoking has no effect on Diabetes prevalence.
* H₁: 🚬 Smoking affects Diabetes prevalence.


In [83]:
table1 = pd.crosstab(df['smoking_status'], df['diabetes'])
table1

diabetes,False,True
smoking_status,,
Former Smoker,78,75
Non-Smoker,83,90
Smoker,97,77


In [84]:
chi2, p_value, dof, expected = chi2_contingency(table1)
print("Chi-square value:", chi2)
print("P-value:", p_value)

print("---------------------------------------------------------------")

if p_value < 0.05:
    print("Reject H0: Smoking status affects diabetes prevalence.")
else:
    print("Accept H0: Smoking status has no significant effect on diabetes.")

Chi-square value: 2.1310933379220915
P-value: 0.3445394542500604
---------------------------------------------------------------
Accept H0: Smoking status has no significant effect on diabetes.


### 🔍 Interpretation:

The p-value (**0.345**) is greater than the significance level (**0.05**), so the difference in diabetes rates across Former Smoker, Non-Smoker, and Smoker groups is **not statistically significant** — it could easily have occurred by chance.

### ✅ Conclusion:

We **accept H₀** — smoking status does not significantly affect diabetes prevalence in this sample. We **cannot conclude** that smoking and diabetes are related here.


#### 🧩 Perform ANOVA Test

In [85]:
daily = df[df['exercise_frequency']=="Daily"]['bmi']
weekly = df[df['exercise_frequency']=="Weekly"]['bmi']
rarely = df[df['exercise_frequency']=="Rarely"]['bmi']
never = df[df['exercise_frequency']=="Never"]['bmi']

In [86]:
groups = [daily, weekly, rarely, never]

k = len(groups)        # number of groups

N = sum(len(group) for group in groups)

print("Number of groups:", k)
print("Total observations:", N)

Number of groups: 4
Total observations: 500


In [87]:
all_bmi = pd.concat(groups)

overall_mean = all_bmi.mean()

print("Overall Mean BMI:", overall_mean)

Overall Mean BMI: 26.5346


In [88]:
SSB = 0

for group in groups:
    group_size = len(group)
    group_mean = group.mean()
    
    SSB += group_size * (group_mean - overall_mean)**2

print("Between Group Sum of Squares:", SSB)



Between Group Sum of Squares: 43.05318482889811


In [89]:
SSW = 0

for group in groups:
    group_mean = group.mean()
    
    SSW += sum((group - group_mean)**2)

print("Within Group Sum of Squares:", SSW)

Within Group Sum of Squares: 11695.3182351711


In [90]:
df_between = k - 1

df_within = N - k

print("DF Between:", df_between)
print("DF Within:", df_within)

DF Between: 3
DF Within: 496


In [91]:
MSB = SSB / df_between

MSW = SSW / df_within


print("Mean Square Between:", MSB)

print("Mean Square Within:", MSW)

Mean Square Between: 14.351061609632703
Mean Square Within: 23.579270635425605


In [92]:
F_value = MSB / MSW

print("F Statistic:", F_value)

F Statistic: 0.608630429309022


In [93]:
p_value = 1 - f.cdf(F_value, df_between, df_within)

print("P-value:", p_value)

alpha = 0.05

print("------------------------------------------------------------")

if p_value < alpha:
    print("Reject H0")
    print("Exercise frequency has a significant effect on BMI.")
    
else:
    print("Accept H0")
    print("Exercise frequency does not have a significant effect on BMI.")

P-value: 0.6096453246623033
------------------------------------------------------------
Accept H0
Exercise frequency does not have a significant effect on BMI.


### 🔍 Interpretation:

The F-statistic (**0.609**) is far below the critical value (**2.623**), and the p-value (**0.610**) is well above **0.05**, meaning exercise frequency explains almost none of the variation in BMI.

### ✅ Conclusion:

We **accept H₀** — exercise frequency (Daily/Weekly/Rarely/Never) does **not** have a significant effect on BMI in this sample.


### 2️⃣ Calculate Confidence Intervals for key numerical data like age, weight, etc.


In [94]:
columns = [
    'age',
    'weight',
    'blood_pressure',
]

In [95]:
confidence = 0.95

for column in columns:
    
    data = df[column]
    
    # Number of observations
    n = len(data)
    
    # Mean
    mean = data.mean()
    
    # Standard deviation
    std = data.std()
    
    # Standard error
    standard_error = std / np.sqrt(n)
    
    # t critical value
    t_value = t.ppf((1 + confidence) / 2, n-1)
    
    # Margin of error
    margin_error = t_value * standard_error
    
    # Confidence interval
    lower_limit = mean - margin_error
    upper_limit = mean + margin_error
    
    print("\nColumn:", column)
    print("Mean:", round(mean,2))
    print("95% Confidence Interval:",
          round(lower_limit,2),
          "to",
          round(upper_limit,2))


Column: age
Mean: 42.67
95% Confidence Interval: 41.09 to 44.26

Column: weight
Mean: 77.51
95% Confidence Interval: 75.8 to 79.21

Column: blood_pressure
Mean: 137.84
95% Confidence Interval: 135.63 to 140.06


### 🔍 Interpretation:

We are **95% confident** the true population mean lies within each interval below:
- 🎂 **Age:** 41.09 to 44.26 years (mean ≈ 42.67)
- ⚖️ **Weight:** 75.8 to 79.21 kg (mean ≈ 77.51)
- ❤️ **Blood Pressure:** 135.63 to 140.06 (mean ≈ 137.84)

### ✅ Conclusion:

All three intervals are **narrow**, so the sample mean estimates are **precise and reliable**. With n = 500, we have good confidence these ranges reflect the true population values.


###  3️⃣ Find the Critical Value and p-value to interpret the test results.


In [96]:
from scipy.stats import f

alpha = 0.05

df_between = 3
df_within = len(df)-4

critical_value = f.ppf(
    1-alpha,
    df_between,
    df_within
)

p_value = 1 - f.cdf(
    F_value,
    df_between,
    df_within
)

print("Critical Value:", critical_value)
print("P-value:", p_value)

print("-------------")

if p_value < alpha:
    print("Reject H0")
else:
    print("Accept H0")

Critical Value: 2.6228791395072832
P-value: 0.6096453246623033
-------------
Accept H0


### 🔍 Interpretation:

This re-confirms the exercise-vs-BMI ANOVA result: F (**0.609**) < critical value (**2.623**), p (**0.610**) > α (**0.05**).

### ✅ Conclusion:

We **accept H₀** — same conclusion as above, exercise frequency has no significant effect on BMI.


### 4️⃣ Perform z-test or t-test based on sample size (mean comparison across groups).


In [97]:
male_bmi = df[df['gender']=="Male"]['bmi']

female_bmi = df[df['gender']=="Female"]['bmi']

In [98]:
n1 = len(male_bmi)
n2 = len(female_bmi)

print("Male sample size:", n1)
print("Female sample size:", n2)

Male sample size: 164
Female sample size: 171


In [99]:
mean1 = male_bmi.mean()

mean2 = female_bmi.mean()

print("Male BMI Mean:", mean1)

print("Female BMI Mean:", mean2)

Male BMI Mean: 26.693902439024384
Female BMI Mean: 26.01988304093567


In [100]:
var1 = male_bmi.var()

var2 = female_bmi.var()

In [101]:
standard_error = np.sqrt(
    (var1/n1) + (var2/n2)
)

In [102]:
t_value = (mean1 - mean2) / standard_error

print("T Statistic:", t_value)

T Statistic: 1.2865443565445367


In [103]:
df_value = n1 + n2 - 2

print("Degrees of Freedom:", df_value)

Degrees of Freedom: 333


In [104]:
p_value = 2 * (1 - t.cdf(abs(t_value), df_value))

print("P-value:", p_value)

P-value: 0.1991468468710984


In [105]:
alpha = 0.05

if p_value < alpha:
    print("Reject H0")
    print("There is a significant difference between male and female BMI.")
    
else:
    print("Accept H0")
    print("There is no significant difference between male and female BMI.")

Accept H0
There is no significant difference between male and female BMI.


### 🔍 Interpretation:

t = **1.287**, df = 333, p-value = **0.199** — well above **0.05**, so the gap between male BMI (26.69) and female BMI (26.02) is **not statistically significant**.

### ✅ Conclusion:

We **accept H₀** — there is **no significant difference** between male and female BMI in this sample.


### 5️⃣ Conduct a chi-square test on categorical data (e.g., 🚬 Smoking habit vs Disease).


In [116]:
table = pd.crosstab(
    df['smoking_status'],
    df['diabetes']
)

table

diabetes,False,True
smoking_status,,
Former Smoker,78,75
Non-Smoker,83,90
Smoker,97,77


In [117]:
row_total = table.sum(axis=1)

column_total = table.sum(axis=0)

total = table.sum().sum()


expected = np.outer(row_total, column_total) / total


expected = pd.DataFrame(
    expected,
    index=table.index,
    columns=table.columns
)


expected

diabetes,False,True
smoking_status,,
Former Smoker,78.948,74.052
Non-Smoker,89.268,83.732
Smoker,89.784,84.216


In [118]:
chi_square = ((table - expected)**2 / expected).sum().sum()

print("Chi-square value:", chi_square)

Chi-square value: 2.1310933379220915


In [119]:
rows = table.shape[0]

columns = table.shape[1]


df_value = (rows-1)*(columns-1)


print("Degrees of Freedom:", df_value)

Degrees of Freedom: 2


In [125]:
p_value = chi2.sf(
    chi_square,
    df_value
)

In [126]:
alpha = 0.05
critical_value = chi2.ppf(
    1-alpha,
    df_value
)


print("Critical Value:", critical_value)

Critical Value: 5.991464547107979


In [127]:
if p_value < alpha:
    
    print("Reject H0")
    print("Smoking status and diabetes are significantly related.")

else:
    
    print("Accept H0")
    print("Smoking status and diabetes are independent.")

Accept H0
Smoking status and diabetes are independent.


### 🔍 Interpretation:

This manual chi-square calculation matches the earlier `chi2_contingency` result (χ² = **2.131**, p = **0.345**), confirming the test is correct.

### ✅ Conclusion:

We **accept H₀** — smoking status and diabetes are **independent**; smoking status does not predict diabetes here.


###  6️⃣ Perform an ANOVA test to check if age groups significantly differ in disease rate.


In [60]:
df['disease_rate'] = df['diabetes'].astype(int)

In [61]:
age_groups = [
    df[df['age_group']=="18-25"]['disease_rate'],
    df[df['age_group']=="26-35"]['disease_rate'],
    df[df['age_group']=="36-45"]['disease_rate'],
    df[df['age_group']=="46-60"]['disease_rate'],
    df[df['age_group']=="60+"]['disease_rate']
]

In [62]:
k = len(age_groups)

N = sum(len(group) for group in age_groups)


print("Number of groups:", k)

print("Total observations:", N)

Number of groups: 5
Total observations: 500


In [63]:
all_data = pd.concat(age_groups)

overall_mean = all_data.mean()


print("Overall Disease Rate:", overall_mean)

Overall Disease Rate: 0.484


In [64]:

SSB = 0

for group in age_groups:
    
    n = len(group)
    
    group_mean = group.mean()
    
    SSB += n * (group_mean - overall_mean)**2


print("SSB:", SSB)

SSB: 1.072553132591968


In [65]:
SSW = 0


for group in age_groups:
    
    group_mean = group.mean()
    
    SSW += sum((group - group_mean)**2)


print("SSW:", SSW)

SSW: 123.79944686740805


In [66]:
df_between = k-1

df_within = N-k


print(df_between)

print(df_within)

4
495


In [67]:
MSB = SSB / df_between

MSW = SSW / df_within


print("MSB:", MSB)

print("MSW:", MSW)

MSB: 0.268138283147992
MSW: 0.2500998926614304


In [68]:
F_value = MSB / MSW


print("F-value:", F_value)

F-value: 1.072124743016107


In [69]:
alpha = 0.05


critical_value = f.ppf(
    1-alpha,
    df_between,
    df_within
)


print("Critical Value:", critical_value)

Critical Value: 2.389947844458197


In [70]:
p_value = 1 - f.cdf(
    F_value,
    df_between,
    df_within
)


print("P-value:", p_value)

P-value: 0.36960744868728046


In [72]:
if p_value < alpha:
    
    print("Reject H0")
    print("Disease rate differs significantly among age groups.")

else:
    
    print("Accept H0")
    print("Disease rate does not differ significantly among age groups.")

Accept H0
Disease rate does not differ significantly among age groups.


### 🔍 Interpretation:

F = **1.072**, p-value = **0.370** — above **0.05**, so diabetes rate does not vary significantly across the five age groups (18-25 through 60+).

### ✅ Conclusion:

We **accept H₀** — age group does **not** significantly affect diabetes rate in this sample, even though age is normally a known risk factor for diabetes in real-world data.


### 7️⃣ Calculate Covariance and Correlation between continuous variables (e.g., Age vs. BMI)


In [73]:
age = df['age']
bmi = df['bmi']


In [74]:
age_mean = age.mean()
bmi_mean = bmi.mean()

print("Mean Age:", age_mean)
print("Mean BMI:", bmi_mean)

Mean Age: 42.674
Mean BMI: 26.534599999999998


In [75]:
n = len(df)


covariance = sum(
    (age - age_mean) *
    (bmi - bmi_mean)) / (n-1)


print("Covariance:", covariance)

Covariance: 2.806893386773546


In [76]:
age_std = age.std()

bmi_std = bmi.std()

In [77]:
correlation = covariance / (age_std * bmi_std)


print("Correlation:", correlation)

Correlation: 0.03211249782651223


### 🔍 Interpretation:

The correlation coefficient (r = **0.032**) is very close to **0**, indicating **almost no linear relationship** between age and BMI.

### ✅ Conclusion:

Age and BMI are **not meaningfully correlated** in this dataset — a person's age gives almost no information about their likely BMI.
